lib/Service/ProvisioningService.php

```diff
@@ -104,15 +106,18 @@ public function getClaimValue(object|array $tokenPayload, string $claimPath, int
 
 		return null;
 	}
+
 	/**
 	 * @param string $tokenUserId
 	 * @param int $providerId
 	 * @param object $idTokenPayload
 	 * @param IUser|null $existingLocalUser
-	 * @return IUser|null
+	 * @return array
 	 * @throws Exception
+	 * @throws PropertyDoesNotExistException
+	 * @throws PreConditionNotMetException
 	 */
-	public function provisionUser(string $tokenUserId, int $providerId, object $idTokenPayload, ?IUser $existingLocalUser = null): ?IUser {
+	public function provisionUser(string $tokenUserId, int $providerId, object $idTokenPayload, ?IUser $existingLocalUser = null): array {
```

Comment:
Not a fan of using arrays that way, but can we annotate the type of the array so psalm can analyze it at least?

Reply:
Thanks, done. I can't predict what's gonna be in `userData` so I keep a simple `array` type. All good?

> Not a fan of using arrays that way

I agree it's not great. The alternative would be to pass the ISession object from the LoginController to `provisioningService->provisionUser()` to be able to change the session in there. Not elegant either.

Reply:
I can also create a `ProvisioningResult` class that contains the user and the raw user data. It's a bit cleaner. Wdyt?